File: src-tauri/src/report.rs
use crate::models::{PeriodStats, RatingBucket, Review};
use chrono::{Days, TimeZone, Utc};
use chrono_tz::Asia::Ho_Chi_Minh;

pub fn today_day_key_vn() -> String {
    Utc::now()
        .with_timezone(&Ho_Chi_Minh)
        .format("%Y-%m-%d")
        .to_string()
}

pub fn yesterday_day_key_vn() -> String {
    let yesterday = Utc::now()
        .with_timezone(&Ho_Chi_Minh)
        .date_naive()
        .checked_sub_days(Days::new(1))
        .unwrap_or_else(|| Utc::now().with_timezone(&Ho_Chi_Minh).date_naive());
    yesterday.format("%Y-%m-%d").to_string()
}

pub fn filter_reviews_by_day(reviews: &[Review], day_key: &str) -> Vec<Review> {
    let mut filtered: Vec<Review> = reviews
        .iter()
        .filter(|review| day_key_vn(review.last_modified_at) == day_key)
        .cloned()
        .collect();
    filtered.sort_by(|a, b| {
        a.star_rating
            .cmp(&b.star_rating)
            .then_with(|| b.last_modified_at.cmp(&a.last_modified_at))
    });
    filtered
}

pub fn build_daily_report_text(day_key: &str, stats: &PeriodStats, reviews: &[Review]) -> String {
    let day_reviews = filter_reviews_by_day(reviews, day_key);
    let review_count = if day_reviews.is_empty() {
        stats.review_count
    } else {
        day_reviews.len() as i64
    };
    let average_rating = if day_reviews.is_empty() {
        stats.average_rating
    } else {
        day_reviews.iter().map(|review| review.star_rating as f64).sum::<f64>() / day_reviews.len() as f64
    };
    let reply_rate = if day_reviews.is_empty() {
        stats.reply_rate
    } else {
        let replied = day_reviews
            .iter()
            .filter(|review| review.has_developer_reply)
            .count() as f64;
        (replied / day_reviews.len() as f64) * 100.0
    };
    let distribution = resolve_rating_distribution(stats, &day_reviews);

    let mut lines = vec![
        format!("📋 Báo cáo MyTV Reviews — {}", format_day_short_vn(day_key)),
        String::new(),
        format!("• Tổng reviews: {}", format_number(review_count)),
        format!("• Rating trung bình: {}★", format_rating(average_rating)),
        format!("• Tỷ lệ đã phản hồi: {}", format_percent(reply_rate)),
        String::new(),
        "Phân bố sao:".to_string(),
    ];

    let mut sorted_distribution = distribution;
    sorted_distribution.sort_by(|a, b| b.stars.cmp(&a.stars));
    for bucket in sorted_distribution {
        lines.push(format!(
            "• {}★: {} ({:.1}%)",
            bucket.stars,
            format_number(bucket.count),
            bucket.percentage
        ));
    }

    lines.push(String::new());
    lines.push(format!(
        "Chi tiết reviews ({}):",
        format_number(day_reviews.len() as i64)
    ));

    if day_reviews.is_empty() {
        lines.push("(Không có review trong ngày này)".to_string());
    } else {
        for (index, review) in day_reviews.iter().enumerate() {
            lines.push(String::new());
            lines.push(format_review_for_report(review, index + 1));
        }
    }

    lines.join("\n")
}

pub fn build_report_subject(day_key: &str) -> String {
    format!("Báo cáo MyTV Reviews — {}", format_day_short_vn(day_key))
}

fn day_key_vn(timestamp: i64) -> String {
    match Ho_Chi_Minh.timestamp_opt(timestamp, 0).single() {
        Some(value) => value.format("%Y-%m-%d").to_string(),
        None => "unknown".to_string(),
    }
}

fn format_day_short_vn(day_key: &str) -> String {
    let parts: Vec<&str> = day_key.split('-').collect();
    if parts.len() != 3 {
        return day_key.to_string();
    }
    format!("{}/{}/{}", parts[2], parts[1], parts[0])
}

fn format_review_for_report(review: &Review, index: usize) -> String {
    let author = review
        .author_name
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .unwrap_or("Ẩn danh");
    let text = review
        .text
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .unwrap_or("(Không có nội dung)")
        .replace('\t', " — ");

    let mut lines = vec![
        format!("{}. {} — {}", index, stars_text(review.star_rating), author),
        format!("   Thời gian: {}", format_timestamp_vn(review.last_modified_at)),
        format!("   Nội dung: {}", text),
    ];

    if review.has_developer_reply {
        let reply = review
            .developer_reply
            .as_deref()
            .map(str::trim)
            .filter(|value| !value.is_empty())
            .unwrap_or("(Không có nội dung phản hồi)");
        lines.push(format!("   Phản hồi: {}", reply));
    } else {
        lines.push("   Phản hồi: Chưa phản hồi".to_string());
    }

    lines.join("\n")
}

fn stars_text(stars: i32) -> String {
    let safe = stars.clamp(1, 5) as usize;
    format!("{}{}", "★".repeat(safe), "☆".repeat(5usize.saturating_sub(safe)))
}

fn distribution_from_reviews(reviews: &[Review]) -> Vec<RatingBucket> {
    let mut counts = [0_i64; 6];
    for review in reviews {
        let stars = review.star_rating.clamp(1, 5) as usize;
        counts[stars] += 1;
    }

    let total = reviews.len() as f64;
    (1..=5)
        .map(|stars| RatingBucket {
            stars,
            count: counts[stars as usize],
            percentage: if total > 0.0 {
                (counts[stars as usize] as f64 / total) * 100.0
            } else {
                0.0
            },
        })
        .collect()
}

fn resolve_rating_distribution(stats: &PeriodStats, day_reviews: &[Review]) -> Vec<RatingBucket> {
    if !day_reviews.is_empty() {
        return distribution_from_reviews(day_reviews);
    }
    if !stats.rating_distribution.is_empty() {
        return stats.rating_distribution.clone();
    }
    (1..=5)
        .map(|stars| RatingBucket {
            stars,
            count: 0,
            percentage: 0.0,
        })
        .collect()
}

fn format_timestamp_vn(timestamp: i64) -> String {
    match Ho_Chi_Minh.timestamp_opt(timestamp, 0).single() {
        Some(value) => value.format("%H:%M %d/%m/%Y").to_string(),
        None => "Không rõ".to_string(),
    }
}

fn format_number(value: i64) -> String {
    let negative = value < 0;
    let digits: Vec<char> = value.abs().to_string().chars().rev().collect();
    let mut parts = String::new();
    for (index, digit) in digits.iter().enumerate() {
        if index > 0 && index % 3 == 0 {
            parts.push('.');
        }
        parts.push(*digit);
    }
    let formatted: String = parts.chars().rev().collect();
    if negative {
        format!("-{formatted}")
    } else {
        formatted
    }
}

fn format_percent(value: f64) -> String {
    format!("{value:.1}%")
}

fn format_rating(value: f64) -> String {
    format!("{value:.2}")
}

File: python/mytv_report/report.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

from .play_api import Review
from .prompt import DEFAULT_DEEPSEEK_EMAIL_SUBJECT

VN_TZ = ZoneInfo("Asia/Ho_Chi_Minh")


@dataclass
class RatingBucket:
    stars: int
    count: int
    percentage: float


@dataclass
class PeriodStats:
    review_count: int
    average_rating: float
    reply_rate: float
    rating_distribution: list[RatingBucket]


def today_day_key_vn() -> str:
    return datetime.now(VN_TZ).strftime("%Y-%m-%d")


def yesterday_day_key_vn() -> str:
    return (datetime.now(VN_TZ).date() - timedelta(days=1)).strftime("%Y-%m-%d")


def day_key_vn(timestamp: int) -> str:
    try:
        return datetime.fromtimestamp(timestamp, tz=VN_TZ).strftime("%Y-%m-%d")
    except (OSError, OverflowError, ValueError):
        return "unknown"


def filter_reviews_by_day(reviews: list[Review], day_key: str) -> list[Review]:
    filtered = [r for r in reviews if day_key_vn(r.last_modified_at) == day_key]
    filtered.sort(key=lambda r: (r.star_rating, -r.last_modified_at))
    return filtered


def build_period_stats(reviews: list[Review]) -> PeriodStats:
    if not reviews:
        return PeriodStats(
            review_count=0,
            average_rating=0.0,
            reply_rate=0.0,
            rating_distribution=_empty_distribution(),
        )

    count = len(reviews)
    avg = sum(r.star_rating for r in reviews) / count
    replied = sum(1 for r in reviews if r.has_developer_reply)
    return PeriodStats(
        review_count=count,
        average_rating=avg,
        reply_rate=(replied / count) * 100.0,
        rating_distribution=distribution_from_reviews(reviews),
    )


def build_daily_breakdown(reviews: list[Review]) -> dict[str, PeriodStats]:
    by_day: dict[str, list[Review]] = defaultdict(list)
    for review in reviews:
        by_day[day_key_vn(review.last_modified_at)].append(review)
    return {day: build_period_stats(items) for day, items in by_day.items()}


def build_daily_report_text(
    day_key: str,
    stats: PeriodStats,
    reviews: list[Review],
) -> str:
    day_reviews = filter_reviews_by_day(reviews, day_key)
    review_count = stats.review_count if not day_reviews else len(day_reviews)
    if day_reviews:
        average_rating = sum(r.star_rating for r in day_reviews) / len(day_reviews)
        replied = sum(1 for r in day_reviews if r.has_developer_reply)
        reply_rate = (replied / len(day_reviews)) * 100.0
        distribution = distribution_from_reviews(day_reviews)
    else:
        average_rating = stats.average_rating
        reply_rate = stats.reply_rate
        distribution = stats.rating_distribution or _empty_distribution()

    lines = [
        f"📋 Báo cáo MyTV Reviews — {format_day_short_vn(day_key)}",
        "",
        f"• Tổng reviews: {_format_number(review_count)}",
        f"• Rating trung bình: {_format_rating(average_rating)}★",
        f"• Tỷ lệ đã phản hồi: {_format_percent(reply_rate)}",
        "",
        "Phân bố sao:",
    ]

    for bucket in sorted(distribution, key=lambda b: b.stars, reverse=True):
        lines.append(
            f"• {bucket.stars}★: {_format_number(bucket.count)} ({bucket.percentage:.1f}%)"
        )

    lines.append("")
    lines.append(f"Chi tiết reviews ({_format_number(len(day_reviews))}):")

    if not day_reviews:
        lines.append("(Không có review trong ngày này)")
    else:
        for index, review in enumerate(day_reviews, start=1):
            lines.append("")
            lines.append(_format_review_for_report(review, index))

    return "\n".join(lines)


def build_zero_review_report_body(day_key: str) -> str:
    return (
        "Kính gửi Anh/Chị,\n\n"
        f"Dưới đây là tóm tắt đánh giá của khách hàng trên Google Play "
        f"trong kỳ báo cáo ngày {day_key}.\n\n"
        "Đánh giá tích cực (0 đánh giá)\n"
        "- Không ghi nhận đánh giá tích cực trong kỳ báo cáo.\n\n"
        "Đánh giá tiêu cực (0 đánh giá)\n"
        "- Không ghi nhận đánh giá tiêu cực trong kỳ báo cáo.\n\n"
        "Trân trọng."
    )


def extract_deepseek_subject(text: str) -> str:
    for line in text.splitlines():
        trimmed = line.strip()
        for prefix in ("Subject:", "Tiêu đề:"):
            if trimmed.startswith(prefix):
                value = trimmed[len(prefix) :].strip()
                if value:
                    return value
    return DEFAULT_DEEPSEEK_EMAIL_SUBJECT


def strip_deepseek_subject_line(text: str) -> str:
    lines = text.splitlines()
    if not lines:
        return text.lstrip()

    first = lines[0].strip()
    if first.startswith("Subject:") or first.startswith("Tiêu đề:"):
        rest = lines[1:]
        while rest and not rest[0].strip():
            rest = rest[1:]
        return "\n".join(rest).lstrip()
    return text.lstrip()


def format_day_short_vn(day_key: str) -> str:
    parts = day_key.split("-")
    if len(parts) != 3:
        return day_key
    return f"{parts[2]}/{parts[1]}/{parts[0]}"


def distribution_from_reviews(reviews: list[Review]) -> list[RatingBucket]:
    counts = [0] * 6
    for review in reviews:
        stars = max(1, min(5, review.star_rating))
        counts[stars] += 1
    total = len(reviews) or 1
    return [
        RatingBucket(
            stars=stars,
            count=counts[stars],
            percentage=(counts[stars] / total) * 100.0 if reviews else 0.0,
        )
        for stars in range(1, 6)
    ]


def _empty_distribution() -> list[RatingBucket]:
    return [RatingBucket(stars=s, count=0, percentage=0.0) for s in range(1, 6)]


def _format_review_for_report(review: Review, index: int) -> str:
    author = (review.author_name or "").strip() or "Ẩn danh"
    text = (review.text or "").strip() or "(Không có nội dung)"
    text = text.replace("\t", " — ")

    lines = [
        f"{index}. {_stars_text(review.star_rating)} — {author}",
        f"   Thời gian: {_format_timestamp_vn(review.last_modified_at)}",
        f"   Nội dung: {text}",
    ]
    if review.has_developer_reply:
        reply = (review.developer_reply or "").strip() or "(Không có nội dung phản hồi)"
        lines.append(f"   Phản hồi: {reply}")
    else:
        lines.append("   Phản hồi: Chưa phản hồi")
    return "\n".join(lines)


def _stars_text(stars: int) -> str:
    safe = max(1, min(5, stars))
    return "★" * safe + "☆" * (5 - safe)


def _format_timestamp_vn(timestamp: int) -> str:
    try:
        return datetime.fromtimestamp(timestamp, tz=VN_TZ).strftime("%H:%M %d/%m/%Y")
    except (OSError, OverflowError, ValueError):
        return "Không rõ"


def _format_number(value: int) -> str:
    negative = value < 0
    digits = list(str(abs(value)))[::-1]
    parts: list[str] = []
    for index, digit in enumerate(digits):
        if index > 0 and index % 3 == 0:
            parts.append(".")
        parts.append(digit)
    formatted = "".join(reversed(parts))
    return f"-{formatted}" if negative else formatted


def _format_percent(value: float) -> str:
    return f"{value:.1f}%"


def _format_rating(value: float) -> str:
    return f"{value:.2f}"
